.github/build_docs.py

```diff
@@ -0,0 +1,303 @@
+    original_dir = pathlib.Path.cwd()
+    try:
+        os.chdir(path)
+        yield
+    finally:
+        os.chdir(original_dir)
+
+
+def find_components_with_docs(repo_root: pathlib.Path) -> List[str]:
+    """
+    Search the repo for component folders that have docs/book.toml.
+
+    Args:
+        repo_root: Repository root directory
+
+    Returns:
+        List of component names with documentation
+    """
+    components = []
+
+    for item in repo_root.iterdir():
+        if not item.is_dir():
+            continue
+
+        docs_path = item / "docs"
+        book_toml_path = docs_path / "book.toml"
+
+        if docs_path.is_dir() and book_toml_path.is_file():
+            components.append(item.name)
+            logger.info(f"Found component with docs: {item.name}")
+
+    return components
+
+
+def generate_api_docs(component_docs_path: pathlib.Path) -> bool:
+    """
+    Generate API documentation using esp-doxybook.
+
+    Args:
+        component_docs_path: Path to component docs directory
+
+    Returns:
+        True if successful, False otherwise
+    """
+    api_md_path = component_docs_path / "src" / "api.md"
+
+    # Create src directory if it doesn't exist
+    (component_docs_path / "src").mkdir(exist_ok=True)
+
+    logger.info(f"Generating API documentation in {api_md_path}")
+
+    try:
+        with change_directory(component_docs_path):
+            # Note: Using relative paths since we're in the docs directory
+            subprocess.run(
+                ["esp-doxybook", "-i", "doxygen_output/xml", "-o", "src/api.md"],
+                check=True,
+                capture_output=True,
+                text=True
+            )
+        logger.info("API documentation generated successfully")
+        return True
+    except subprocess.CalledProcessError as e:
+        logger.warning(f"Failed to generate API documentation: {e}")
+        logger.warning(f"stdout: {e.stdout if hasattr(e, 'stdout') else 'N/A'}")
+        logger.warning(f"stderr: {e.stderr if hasattr(e, 'stderr') else 'N/A'}")
+        logger.warning("Continuing with mdbook build...")
+        return False
+
+
+def build_component_docs(component_name: str, config: BuildConfig) -> bool:
+    """
+    Build documentation for a component using mdbook.
+
+    Args:
+        component_name: Component name
+        config: Build configuration
+
+    Returns:
+        True if build was successful, False otherwise
+    """
+    component_docs_path = config.repo_root / component_name / "docs"
+
+    logger.info(f"Building docs for {component_name}...")
+
+    try:
+        # Generate API documentation
+        generate_api_docs(component_docs_path)
+
+        # Build mdbook
+        env = os.environ.copy()
+
+        # Set site-url based on whether this is a PR or merge event
+        if config.pr_number:
+            site_url = f"/idf-extra-components/pr-preview-{config.pr_number}/{component_name}/"
```

Comment:
If I understand it right, you are running the docs build for PRs as well.

When the docs are built for PR, do we somehow preserve the content of the /latest/ directory?

Reply:
Hi @igrr yes, we only run the docs build for PR at the moment. But there's no deploy for the PR build. And there's also no docs preview. (left for future when we think it's necessary). The "Prepare files for deployment 📁" and "Deploy to Github Pages 🚀" only available for a "Push to Master" event.